src/Bicep.Core/Semantics/SemanticModel.cs

```diff
@@ -579,6 +631,45 @@ private IEnumerable<IDiagnostic> GatherParameterMismatchDiagnostics(ISemanticMod
             }
         }
 
+        private IEnumerable<IDiagnostic> GatherMissingRequiredExtensionConfigAssignmentDiagnostics(ISemanticModel usingModel)
+        {
+            if (!usingModel.Features.ModuleExtensionConfigsEnabled)
+            {
+                yield break;
+            }
+
+            // emit diagnostic only if there is a using statement
+            var usingDeclarationSyntax = this.Root.UsingDeclarationSyntax;
+
+            if (usingDeclarationSyntax?.Path is NoneLiteralSyntax)
+            {
+                yield break;
+            }
+
+            // assignment symbols that do not match to an extension in the main file
+            var assignmentAliasesWithMissingExtension = Root.ExtensionConfigAssignments.Where(a => TryGetExtensionMetadata(a) is null);
+
+            // assignments that are missing
+            var missingRequiredAssignments = usingModel.Extensions
+                .Where(kvp => kvp.Value.RequiresConfigAssignment && TryGetExtensionConfigAssignment(kvp.Value) is null)
+                .Select(kvp => (kvp.Key, kvp.Value.ConfigAssignmentDeclaredType!))
+                .OrderBy(kvp => kvp.Key)
+                .ToImmutableArray();
```

Comment:
nit: We can use `ToArray()` here instead, since `missingRequiredAssignments` is short-lived and this avoids the overhead of `ToImmutableArray()`.